hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/RandomGeneratePrecompiledContract.java

```diff
@@ -0,0 +1,136 @@
+import org.hyperledger.besu.evm.gascalculator.GasCalculator;
+import org.hyperledger.besu.evm.precompile.AbstractPrecompiledContract;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.util.concurrent.ExecutionException;
+import java.util.function.Supplier;
+
+import static com.hedera.services.exceptions.ValidationUtils.validateTrue;
+import static com.hedera.services.txns.util.RandomGenerateTransitionLogic.randomNumFromBytes;
+
+/**
+ * System contract to generate random numbers. This will generate 256-bit pseudorandom number when no range is provided.
+ * If a given 32-bit range is provided returns a pseudorandom 32-bit integer X belonging to [0, range).
+ * The pseudorandom number is generated using n-3 record's running hash.
+ */
+@Singleton
+public class RandomGeneratePrecompiledContract extends AbstractPrecompiledContract {
+	private static final String PRECOMPILE_NAME = "RandomGenerate";
+	private static final LongType WORD_DECODER = TypeFactory.create("uint32");
+	//random256BitGenerator(uint256)
+	static final int RANDOM_256_BIT_GENERATOR_SELECTOR = 0x267dc6a3;
+	//randomNumberGeneratorInRange(uint32)
+	static final int RANDOM_NUM_IN_RANGE_GENERATOR_SELECTOR = 0x85b4610c;
+	public static final String RANDOM_GENERATE_PRECOMPILE_ADDRESS = "0x169";
+
+	private final GlobalDynamicProperties dynamicProperties;
+	private final Supplier<RecordsRunningHashLeaf> runningHashLeafSupplier;
+
+	@Inject
+	public RandomGeneratePrecompiledContract(final GasCalculator gasCalculator,
+			final GlobalDynamicProperties dynamicProperties,
+			final Supplier<RecordsRunningHashLeaf> runningHashLeafSupplier) {
+		super(PRECOMPILE_NAME, gasCalculator);
+		this.dynamicProperties = dynamicProperties;
+		this.runningHashLeafSupplier = runningHashLeafSupplier;
+	}
+
+	@Override
+	public long gasRequirement(final Bytes bytes) {
+		return dynamicProperties.randomGenerateGasCost();
+	}
+
+	@Override
+	public Bytes compute(final Bytes input, final MessageFrame frame) {
+		try {
+			final var selector = input.getInt(0);
+			return switch (selector) {
+				case RANDOM_256_BIT_GENERATOR_SELECTOR -> random256BitGenerator();
+				case RANDOM_NUM_IN_RANGE_GENERATOR_SELECTOR -> randomNumberGeneratorInRange(input);
+				default -> null;
+			};
+		} catch (IndexOutOfBoundsException ignore) {
+			return null; // should we ignore all exceptions and return null ?
+		}
+	}
+
+	private Bytes randomNumberGeneratorInRange(final Bytes input) {
+		final var range = rangeValueFrom(input);
+		// if range is invalid fail here
+		validateTrue(range >= 0, ResponseCodeEnum.INVALID_RANDOM_GENERATE_RANGE);
+
+		final var randomNum = generateRandomNumberFromHash(range);
+		// if hash is invalid or not present returns null
+		if (randomNum == -1) {
+			return null;
+		}
+		return padded(randomNum);
+	}
+
+	private int generateRandomNumberFromHash(final int range) {
+		final var hash = getHash();
+		if (hash == null) {
+			return -1;
```

Comment:
Since this method returns a random number primary `int` returned -1 . `-1` sentinel is returned because random number will be generated from `0 <= X < range` and can't be negative.